File: libs-node/queueClient.js
"use strict";
/*jshint unused:false*/
var sqs = require("./aws").SQS;
var config = require('config').config;
var MongoClient = require('mongodb').MongoClient;

function AWSQueueClient() {

}

AWSQueueClient.prototype.sendMessage = function(messageBody, queueUrl, next) {
    var message = {
        MessageBody: messageBody,
        QueueUrl: queueUrl
    };
    sqs.sendMessage(message, function(err) {
        if (err) {
            next(err);
        } else {
            next();
        }
    });
};

function LocalQueueClient() {
    var self = this;
}

LocalQueueClient.prototype.runQuery = function(next) {
    if (self.db === undefined) {
        MongoClient.connect(config.MONGODB_INSTANCE + "/" + config.MONGODB_QUEUE_DB, function(err, db) {
            if (!err) {
                self.db = db;
                next();
            } else {
                console.log(err);
            }
        });
    } else {
        next();
    }
};


LocalQueueClient.prototype.sendMessage = function(msg, queue, next) {
    self.runQuery(function(msg, queue, next) {
        self.db.collection(queue).insert({
            'inProg': false,
            'done': false,
            'msg': new Buffer(msg).toString('base64')
        }, function(err){
            if (!err) {
                next();
            } else {
                console.log(err);
            }
        });
    });
};

function QueueClient(provider) {
    if (provider === "AWS") {
        return new AWSQueueClient();
    } else if (provider === "LOCAL") {
        return new LocalQueueClient();
    }
}

module.exports = QueueClient;

File: lwutils/queue/queue_client.py
from lwutils.queue.clients.aws_queue_client import AWSQueueClient
from lwutils.queue.clients.local_queue_client import LocalQueueClient
from lwutils.queue.clients.inmemory_queue_client import InMemoryQueueClient


class QueueClient(object):

    def factory(config):
        if config.get('QUEUE_PROVIDER') == "AWS":
            return AWSQueueClient(config)
        elif config.get('QUEUE_PROVIDER') == "LOCAL":
            return LocalQueueClient(config)
        elif config.get('QUEUE_PROVIDER') == "INMEMORY":
            return InMemoryQueueClient(config)
        else:
            print "Unknown Queue provider: " + config.get('QUEUE_PROVIDER', "not configured")
    factory = staticmethod(factory)
